_source/_posts/2022-08-19-build-crud-spring-and-vue.md

```diff
@@ -0,0 +1,1057 @@
+
+## Test the unsecured app and secure it
+
+Run the app using the following command from the `resource-server` subdirectory.
+
+```bash
+./gradlew bootRun
+```
+
+Open a new Bash shell and use HTTPie to test the resource server.
+
+```bash
+http :9000/todos
+```
+
+```bash
+HTTP/1.1 200 
+...
+
+{
+    "_embedded": {
+        "todos": [
+            {
+                "_links": {
+                    "self": {
+                        "href": "http://localhost:9000/todos/1"
+                    },
+                    "todo": {
+                        "href": "http://localhost:9000/todos/1"
+                    }
+                },
+                "completed": false,
+                "id": 1,
+                "title": "Buy milk"
+            },
+            {
+                "_links": {
+                    "self": {
+                        "href": "http://localhost:9000/todos/2"
+                    },
+                    "todo": {
+                        "href": "http://localhost:9000/todos/2"
+                    }
+                },
+                "completed": true,
+                "id": 2,
+                "title": "Eat pizza"
+            },
+            ...
+        ]
+    },
+...
+}
+```
+
+Stop the resource server using `control-c`. 
+
+Edit the security configuration file, updating the security configuration bean definition to the following.
+
+`src/main/java/com/example/demo/OAuth2ResourceServerSecurityConfiguration.java`
+
+```java
+@Bean
+public SecurityFilterChain filterChain(HttpSecurity http) throws Exception {
+    http.authorizeRequests()
+            .anyRequest().authenticated()
+            .and()
+            .oauth2ResourceServer().jwt();
+    return http.build();
+}
+```
+
+This configuration requires JWT auth on all requests.
+
+Restart the server. Use `control-c` to stop it if it's running.
+
+```bash
+./gradlew bootRun
+```
+
+Use HTTPie again to try and request the todo items.
+
+```bash
+http :9000/todos
+```
+
+You will get an error.
+
+```bash
+HTTP/1.1 401 
+...
+
+401 Unauthorized
+
+```
+
+The resource server is finished. The next step is to create the Vue client.
+
+## Create the Vue JavaScript client
+
+From the project parent directory, use the Vue CLI to create a new application and navigate into the newly created `client` directory.
+
+```bash
+vue create client
+cd client
+```
+
+Pick `Default Vue 3`
+
+Wait for it to finish. Add the Quasar framework.
+
+```bash
+vue add quasar
+```
+
+You can just accept the defaults. For me they were the following.
+
+- **Allow Quasar to replace App.vue, About.vue, Home.vue and (if available) router.js?** Yes
+- **Pick your favorite CSS preprocessor:** Sass
+- **Choose Quasar Icon Set:** Material
+- **Default Quasar language pack:** en-US
+- **Use RTL support?** No
+
+Add our dependencies.
+
+```bash
+npm i axios@^0.27.2 vuejs3-logger@^1.0.0 vue-router@^4.0.16 @okta/okta-vue@^5.3.0
```

Comment:
We typically lock versions (with no ranges) since so many npm libraries break between minor releases. However, developers often ignore what we say and trim off the versions. You can do what you'd normally recommend here.

Reply:
I get an error when I run this:

```
npm i axios@^0.27.2 vuejs3-logger@^1.0.0 vue-router@^4.0.16 @okta/okta-vue@^5.3.0
```

If I remove the ^, it works. 

```
npm i axios@0.27.2 vuejs3-logger@1.0.0 vue-router@4.0.16 @okta/okta-vue@5.3.0
```